pkg/monitoring/vmstat/hyperv/hyperv_wmi.go

```diff
@@ -0,0 +1,234 @@
+type HealthState uint16
+type Heartbeat uint16
+
+type msvm_ComputerSystem struct {
+	ElementName           string
+	InstallDate           string
+	HealthState           HealthState
+	EnabledState          EnabledState
+	OtherEnabledState     string
+	RequestedState        EnabledState
+	TimeOfLastStateChange string
+	Name                  string
+	ProcessID             uint32
+}
+
+// some of fields are pointers instead of values as in some cases they may not exists
+// during WMI query. For example if VM is not running MemoryUsage, ProcessorLoad and Heartbeat
+// are not available
+type msvm_SummaryInformation struct {
+	Name                 string
+	ElementName          string
+	GuestOperatingSystem *string
+	Version              string
+	EnabledState         EnabledState
+	HealthState          HealthState
+	CreationTime         time.Time
+	// MemoryUsage is in megabytes
+	MemoryUsage        *uint64
+	Uptime             uint64
+	ProcessorLoad      *uint16
+	Heartbeat          *Heartbeat
+	NumberOfProcessors uint16
+}
+
+type msvm_GuestNetworkAdapterConfiguration struct {
+	InstanceID  string
+	IPAddresses []string
+}
+
+type hypervStat struct {
+	TotalCPUWaitTimePerDispatch float64
+	List                        map[string]vmStat
+}
+
+var hypervPath = "\\Hyper-V Hypervisor Virtual processor(*)\\CPU Wait Time Per Dispatch"
+
+func (im *impl) GetMeasurements() (map[string]interface{}, error) {
+	meas := make(map[string]interface{})
+
+	stat := &hypervStat{
+		List: make(map[string]vmStat),
+	}
+
+	var countersErr error
+	res, countersErr := im.watcher.GetFormattedQueryData(hypervPath)
+	if countersErr == nil {
+		for _, c := range res {
+			switch c.InstanceName {
+			case "_Total":
+				meas["cpu_wait_time_per_dispatch_total_ns"] = int64(c.Value)
+			default:
+				name := strings.Split(c.InstanceName, ":")[0]
+				stat.List[name] = vmStat{
+					InstanceName:           name,
+					CPUWaitTimePerDispatch: c.Value,
+				}
+			}
+		}
+	} else {
+		log.WithError(countersErr)
+	}
+
+	var list []interface{}
+
+	ipMap := make(map[string][]string)
+
+	var q string
+	var ips []msvm_GuestNetworkAdapterConfiguration
+	q = wmi.CreateQuery(&ips, "")
+	if err := wmi.QueryNamespace(q, &ips, `root\virtualization\v2`); err == nil {
+		for i := range ips {
+			// InstanceID is in form of Microsoft:GuestNetwork\GUID\GUID
+			// First GUID matches field Name in struct msvm_SummaryInformation
+			if !guestNetworkRegexp.Match([]byte(ips[i].InstanceID)) {
+				log.Warn("invalid InstanceID in hyper-v response. expected form:"+
+					"[Microsoft:GuestNetwork\\<GUID>\\<GUID>] actual form: [%s]", ips[i].InstanceID)
+			} else {
+				id := guestNetworkRegexp.FindAllStringSubmatch(ips[i].InstanceID, -1)
+				ipMap[id[0][1]] = ips[i].IPAddresses
+			}
+		}
+	} else {
+		log.WithError(err)
+	}
+
+	var dst []msvm_SummaryInformation
+	q = wmi.CreateQuery(&dst, "")
+
+	if err := wmi.QueryNamespace(q, &dst, `root\virtualization\v2`); err != nil {
+		log.WithError(err)
```

Comment:
the same here. `log.WithError(err)` doesn't log the error

Reply:
done